liquid-staking/src/lib.rs

```diff
@@ -255,78 +299,49 @@ pub trait LiquidStaking<ContractReader>:
-        let unstake_amount = unstake_token_attributes.unstake_amount;
-        let delegation_contract_mapper = self.delegation_contract_data(&delegation_contract);
-        let delegation_contract_data = delegation_contract_mapper.get();
-        if delegation_contract_data.total_unbonded_from_ls_contract >= unstake_amount {
-            delegation_contract_mapper.update(|contract_data| {
-                contract_data.total_unstaked_from_ls_contract -= &unstake_amount;
-                contract_data.total_unbonded_from_ls_contract -= &unstake_amount
-            });
+        let delegation_contract = unstake_token_attributes.delegation_contract.clone();
 
-            storage_cache.total_withdrawn_egld -= &unstake_amount;
-            self.unstake_token_supply()
-                .update(|x| *x -= &unstake_amount);
-            self.burn_unstake_tokens(payment.token_nonce);
-            self.send().direct_egld(&caller, &unstake_amount)
-        } else {
-            let gas_for_async_call = self.get_gas_for_async_call();
-            self.delegation_proxy_obj()
-                .contract(delegation_contract.clone())
-                .withdraw()
-                .with_gas_limit(gas_for_async_call)
-                .async_call()
-                .with_callback(LiquidStaking::callbacks(self).withdraw_tokens_callback(
-                    caller,
-                    delegation_contract,
-                    payment.token_nonce,
-                    unstake_amount,
-                ))
-                .call_and_exit();
-        }
+        let gas_for_async_call = self.get_gas_for_async_call();
+
+        drop(storage_cache);
+        self.delegation_proxy_obj()
+            .contract(delegation_contract)
+            .withdraw()
+            .with_gas_limit(gas_for_async_call)
+            .async_call()
+            .with_callback(LiquidStaking::callbacks(self).withdraw_tokens_callback(
+                unstake_token_attributes,
+                caller,
+                payment.token_nonce,
+            ))
+            .call_and_exit();
     }
 
     #[callback]
     fn withdraw_tokens_callback(
         &self,
+        unstake_token_attributes: UnstakeTokenAttributes<Self::Api>,
         caller: ManagedAddress,
-        delegation_contract: ManagedAddress,
-        unstake_token_nonce: u64,
-        unstake_token_amount: BigUint,
+        nonce_to_burn: u64,
         #[call_result] result: ManagedAsyncCallResult<()>,
     ) {
         match result {
             ManagedAsyncCallResult::Ok(()) => {
                 let withdraw_amount = self.call_value().egld_value().clone_value();
                 let mut storage_cache = StorageCache::new(self);
                 let delegation_contract_mapper =
-                    self.delegation_contract_data(&delegation_contract);
+                    self.delegation_contract_data(&unstake_token_attributes.delegation_contract);
                 if withdraw_amount > 0u64 {
```

Comment:
It is important to keep in mind that the withdraw process happens at the SC's level, and not for a specific user. I think we should only update the storage (total_withdrawn_egld and total_unbonded_from_ls_contract) inside the if statement withdraw_amount > 0u64. In other words, we should update unbond_from_provider and burn the tokens, regardless the withdraw amount (as the tokens may have been previously withdrawn by other user).
Moreover, maybe we would be more safe if we cover the case of an error as well (by sending back the tokens to the user if the withdraw callback receives an error).